drivers/counter/counter_mchp_xec.c

```diff
@@ -0,0 +1,413 @@
+				     const struct counter_top_cfg *cfg)
+{
+	volatile struct counter_xec_registers *counter =
+						COUNTER_XEC_REG_BASE(dev);
+	const struct counter_xec_config *counter_cfg = COUNTER_XEC_CONFIG(dev);
+	struct counter_xec_data *data = COUNTER_XEC_DATA(dev);
+
+	if (counter->control_reg.start || data->alarm_cb) {
+		return -EBUSY;
+	}
+
+	if (cfg->ticks != counter_cfg->info.max_top_value) {
+		return -ENOTSUP;
+	}
+
+	if (counter_cfg->info.max_top_value == UINT16_MAX) {
+		counter->preload16_reg = cfg->ticks;
+	} else {
+		counter->preload32_reg = cfg->ticks;
+	}
+
+	counter->count_reg = cfg->ticks;
+
+	data->top_cb = cfg->callback;
+	data->user_data = cfg->user_data;
+
+	if (data->top_cb) {
+		counter->int_enable_reg.int_enable = 1;
+		counter->control_reg.auto_restart = 1;
+	} else {
+		counter->int_enable_reg.int_enable = 0;
+		counter->control_reg.auto_restart = 0;
+	}
+
+	LOG_DBG("Counter top value was set to %u", cfg->ticks);
+
+	return 0;
+}
+
+static u32_t counter_xec_get_max_relative_alarm(struct device *dev)
+{
+	const struct counter_xec_config *counter_cfg = COUNTER_XEC_CONFIG(dev);
+
+	return counter_cfg->info.max_top_value;
+}
+
+static void counter_xec_isr(struct device *dev)
+{
+	volatile struct counter_xec_registers *counter =
+						COUNTER_XEC_REG_BASE(dev);
+	const struct counter_xec_config *counter_cfg = COUNTER_XEC_CONFIG(dev);
+	struct counter_xec_data *data = COUNTER_XEC_DATA(dev);
+	u32_t girq_result;
+
+	girq_result = MCHP_GIRQ_RESULT(counter_cfg->girq_id);
+	REG32(MCHP_GIRQ_SRC_ADDR(counter_cfg->girq_id)) = girq_result;
+
+	if (!(girq_result & counter_cfg->girq_src)) {
+		return;
+	}
+
+	LOG_DBG("Counter ISR");
+
+	if (data->alarm_cb) {
+		u32_t ticks;
+
+		if (counter_cfg->info.max_top_value == UINT16_MAX) {
+			ticks = counter->preload16_reg;
+		} else {
+			ticks = counter->preload32_reg;
+		}
+
+		data->alarm_cb(dev, 0, ticks, data->user_data);
+	} else if (data->top_cb) {
+		data->top_cb(dev, data->user_data);
+	}
+}
+
+static const struct counter_driver_api counter_xec_api = {
+		.start = counter_xec_start,
+		.stop = counter_xec_stop,
+		.read = counter_xec_read,
+		.set_alarm = counter_xec_set_alarm,
+		.cancel_alarm = counter_xec_cancel_alarm,
+		.set_top_value = counter_xec_set_top_value,
+		.get_pending_int = counter_xec_get_pending_int,
+		.get_top_value = counter_xec_get_top_value,
+		.get_max_relative_alarm = counter_xec_get_max_relative_alarm,
+};
+
+static int counter_xec_init(struct device *dev)
+{
+	volatile struct counter_xec_registers *counter =
+						COUNTER_XEC_REG_BASE(dev);
+	const struct counter_xec_config *counter_cfg = COUNTER_XEC_CONFIG(dev);
+
+	counter_xec_stop(dev);
+
+	counter->control_reg.count_up     = 0;
+	counter->control_reg.pre_scale    = counter_cfg->prescaler;
+
+	MCHP_GIRQ_ENSET(counter_cfg->girq_id) = counter_cfg->girq_src;
+	MCHP_GIRQ_BLK_SETEN(counter_cfg->girq_id);
+
+	counter_cfg->config_func();
+
+	return 0;
+}
+
+#if defined(DT_MICROCHIP_XEC_TIMER_0)
```

Comment:
why not create macro that will take instance index. Less code repetition. Something like:
https://github.com/zephyrproject-rtos/zephyr/blob/77db273f6f84b3d94d34b5e72070ec6bc31e41cf/drivers/counter/counter_nrfx_rtc.c#L332

Reply:
It's a preference issue, why not.

Just waiting for the day we will instantiate automatically via dts based auto generation.